ngsderive/commands/instrument.py

```diff
@@ -69,42 +70,51 @@
     "^H[A-Z0-9]{4}ALXX$": ["HiSeq X"],  # (8-lane) flow cell
     "^H[A-Z0-9]{4}BGX[A-Z,0-9]$": ["NextSeq"],  # High output flow cell
     "^H[A-Z0-9]{4}AFXX$": ["NextSeq"],  # Mid output flow cell
-    "^H[A-Z0-9]{5}RXX$": ["NovaSeq"],  # S1 flow cell
-    "^H[A-Z0-9]{5}RXX$": ["NovaSeq"],  # SP flow cell
-    "^H[A-Z0-9]{5}MXX$": ["NovaSeq"],  # S2 flow cell
-    "^H[A-Z0-9]{5}SXX$": ["NovaSeq"],  # S4 flow cell
+    "^H[A-Z0-9]{5}RXX$": ["NovaSeq 6000"],  # S1 flow cell
+    "^H[A-Z0-9]{5}RXX$": ["NovaSeq 6000"],  # SP flow cell
+    "^H[A-Z0-9]{5}MXX$": ["NovaSeq 6000"],  # S2 flow cell
+    "^H[A-Z0-9]{5}SXX$": ["NovaSeq 6000"],  # S4 flow cell
     "^A[A-Z0-9]{4}$": ["MiSeq"],  # MiSeq flow cell
     "^B[A-Z0-9]{4}$": ["MiSeq"],  # MiSeq flow cell
     "^D[A-Z0-9]{4}$": ["MiSeq"],  # MiSeq nano flow cell
     # "^D[A-Z0-9]{4}$" : ["HiSeq 2000", "HiSeq 2500"],  # Unknown HiSeq flow cell examined in SJ data
     "^G[A-Z0-9]{4}$": ["MiSeq"],  # MiSeq micro flow cell
 }
 
+# There is an upgrade kit for HiSeq 2000 machines which makes them indistinguishable
+# from Hiseq 2500 machines. Since we can't know whether an upgrade kit has been
+# applied, we can never differentiate between HiSeq 2000 and 2500 machines.
 upgrade_sets = [(set(["HiSeq 2000", "HiSeq 2500"]), ["HiSeq 2000", "HiSeq 2500"])]
 
+# if item[0] pattern matches, remove item[1] from set, add item[2] to set
+specific_iid_patterns = [
+    ("^K00(?:258|3(?:09|2[679]))$", set(["HiSeq 3000"]), set(["HiSeq 4000"])),
+    ("^NB551266$", set(["NextSeq"]), set(["NextSeq 550"])),
+    ("^NS500229$", set(["NextSeq"]), set(["NextSeq 500"])),
+]
+
 
 def derive_instrument_from_iid(iid):
     matching_instruments = set()
 
     for pattern in instrument_ids.keys():
         if re.search(pattern, iid):
             matching_instruments |= set(instrument_ids[pattern])
+    for pattern, set_to_remove, set_to_add in specific_iid_patterns:
+        if re.search(pattern, iid):
+            matching_instruments.difference_update(set_to_remove)
+            matching_instruments.update(set_to_add)
 
     return matching_instruments
 
 
 def predict_instrument_from_iids(iids):
     possible_instruments_by_iid = set()
-    detected_at_least_one_instrument = False
 
     for iid in iids:
-        if len(possible_instruments_by_iid) == 0:
-            possible_instruments_by_iid = derive_instrument_from_iid(iid)
-            detected_at_least_one_instrument = True
-        else:
-            possible_instruments_by_iid &= derive_instrument_from_iid(iid)
+        possible_instruments_by_iid |= derive_instrument_from_iid(iid)
 
-    return possible_instruments_by_iid, detected_at_least_one_instrument
+    return possible_instruments_by_iid
```

Comment:
If you remove this conditional, how do you know know the difference between (a) no instruments were predicted from iids, leaving you with an empty set and (b) conflicting instruments were predicted from iids, leaving you with an empty set.

Reply:
Take a closer look at what happens this time this code runs. `detected_at_least_one_instrument` will _ALWAYS_ be set to `True` on the first pass, because `possible_instruments_by_iid` is initialed with `len() == 0`.

I also believe these loops should be _building_ a list of a possible instruments via the `union` operator. This block of logic shouldn't be in the business of detecting conflicting instruments. That part of the logic is appropriately handled by the `resolve_instrument()` function.